In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity

In [2]:
df = pd.read_excel('Assignment 5.xls', index_col='User')
ndf = pd.read_excel('Assignment 5.xls', index_col='User', sheetname='NormRatings')

In [3]:
df.drop('Mean', inplace=True, axis=1) # remove coluna Mean
l2_norm = df.loc['L2'] # salva l2 norm
df.drop('L2', inplace=True) # remove row l2 norm

ndf.drop('L2', inplace=True) # remove row l2 norm

    Similarity between Toy Story and Star Wars, raw: 0.645
    Similarity between Toy Story and Star Wars, normalized: -0.378

In [4]:
df_cos = cosine_similarity(df.T.fillna(0), df.T.fillna(0))
dfcos = pd.DataFrame(df_cos, index=df.columns, columns=df.columns)

dfcos['1: Toy Story (1995)']

1: Toy Story (1995)                                        1.000000
1210: Star Wars: Episode VI - Return of the Jedi (1983)    0.644995
356: Forrest Gump (1994)                                   0.580540
318: Shawshank Redemption, The (1994)                      0.667424
593: Silence of the Lambs, The (1991)                      0.570229
3578: Gladiator (2000)                                     0.587852
260: Star Wars: Episode IV - A New Hope (1977)             0.747409
2028: Saving Private Ryan (1998)                           0.534579
296: Pulp Fiction (1994)                                   0.667846
1259: Stand by Me (1986)                                   0.492659
2396: Shakespeare in Love (1998)                           0.376659
2916: Total Recall (1990)                                  0.623056
780: Independence Day (ID4) (1996)                         0.690665
541: Blade Runner (1982)                                   0.383067
1265: Groundhog Day (1993)                      

In [5]:
ndf_cos = cosine_similarity(ndf.T.fillna(0), ndf.T.fillna(0))
ndfcos = pd.DataFrame(ndf_cos, index=ndf.columns, columns=ndf.columns)

ndfcos['1: Toy Story (1995)']

1: Toy Story (1995)                                        1.000000
1210: Star Wars: Episode VI - Return of the Jedi (1983)   -0.377771
356: Forrest Gump (1994)                                   0.355780
318: Shawshank Redemption, The (1994)                      0.215975
593: Silence of the Lambs, The (1991)                     -0.603708
3578: Gladiator (2000)                                     0.139214
260: Star Wars: Episode IV - A New Hope (1977)            -0.050338
2028: Saving Private Ryan (1998)                           0.192799
296: Pulp Fiction (1994)                                   0.295013
1259: Stand by Me (1986)                                  -0.225657
2396: Shakespeare in Love (1998)                           0.116755
2916: Total Recall (1990)                                  0.184712
780: Independence Day (ID4) (1996)                        -0.319904
541: Blade Runner (1982)                                   0.020007
1265: Groundhog Day (1993)                      

# Top 5 Toy Story

## Unnormalized

In [6]:
dfcos['1: Toy Story (1995)'].nlargest(6)

1: Toy Story (1995)                               1.000000
260: Star Wars: Episode IV - A New Hope (1977)    0.747409
780: Independence Day (ID4) (1996)                0.690665
296: Pulp Fiction (1994)                          0.667846
318: Shawshank Redemption, The (1994)             0.667424
1265: Groundhog Day (1993)                        0.661016
Name: 1: Toy Story (1995), dtype: float64

## Normalized

In [7]:
ndfcos['1: Toy Story (1995)'].nlargest(6)

1: Toy Story (1995)                      1.000000
34: Babe (1995)                          0.554448
356: Forrest Gump (1994)                 0.355780
296: Pulp Fiction (1994)                 0.295013
318: Shawshank Redemption, The (1994)    0.215975
2028: Saving Private Ryan (1998)         0.192799
Name: 1: Toy Story (1995), dtype: float64

# Top 5 for User 5277

In [8]:
df.loc[5277]

1: Toy Story (1995)                                        1.0
1210: Star Wars: Episode VI - Return of the Jedi (1983)    NaN
356: Forrest Gump (1994)                                   NaN
318: Shawshank Redemption, The (1994)                      2.0
593: Silence of the Lambs, The (1991)                      4.0
3578: Gladiator (2000)                                     2.0
260: Star Wars: Episode IV - A New Hope (1977)             5.0
2028: Saving Private Ryan (1998)                           NaN
296: Pulp Fiction (1994)                                   NaN
1259: Stand by Me (1986)                                   4.0
2396: Shakespeare in Love (1998)                           3.0
2916: Total Recall (1990)                                  2.0
780: Independence Day (ID4) (1996)                         2.0
541: Blade Runner (1982)                                   NaN
1265: Groundhog Day (1993)                                 2.0
2571: Matrix, The (1999)                               

nota do usuário * similaridade dos items / sum(similaridades)

## Unnormalized

In [14]:
numerator = df.loc[5277].fillna(0).dot(dfcos)
denominator = df.loc[5277].notnull().dot(dfcos)

(numerator / denominator).nlargest(5)

527: Schindler's List (1993)                      2.973883
1259: Stand by Me (1986)                          2.928801
260: Star Wars: Episode IV - A New Hope (1977)    2.922240
593: Silence of the Lambs, The (1991)             2.883304
2396: Shakespeare in Love (1998)                  2.852131
Name: 5277, dtype: float64

In [31]:
numerator = df.loc[3048].fillna(0).dot(dfcos)
denominator = df.loc[3048].notnull().dot(dfcos)

(numerator / denominator).nlargest(5)

296: Pulp Fiction (1994)                                   3.420229
318: Shawshank Redemption, The (1994)                      3.405435
34: Babe (1995)                                            3.368467
1210: Star Wars: Episode VI - Return of the Jedi (1983)    3.268356
1: Toy Story (1995)                                        3.265483
Name: 3048, dtype: float64

In [34]:
denominator.to_dict()

{'1198: Raiders of the Lost Ark (1981)': 6.5109526268931139,
 '1210: Star Wars: Episode VI - Return of the Jedi (1983)': 7.3577007648527539,
 '1259: Stand by Me (1986)': 6.0597760845723183,
 '1265: Groundhog Day (1993)': 6.4673671177044261,
 '1: Toy Story (1995)': 7.8197028709620664,
 '2028: Saving Private Ryan (1998)': 6.6513423029178869,
 '2396: Shakespeare in Love (1998)': 5.8083671587818904,
 '2571: Matrix, The (1999)': 5.5135229642438341,
 '260: Star Wars: Episode IV - A New Hope (1977)': 8.0589125012704965,
 '2762: Sixth Sense, The (1999)': 7.1286952347917776,
 '2916: Total Recall (1990)': 5.9711879911819583,
 '296: Pulp Fiction (1994)': 7.0773467531528915,
 '318: Shawshank Redemption, The (1994)': 6.8212253352570613,
 '34: Babe (1995)': 6.4802403870320386,
 '356: Forrest Gump (1994)': 6.5381041752776481,
 '3578: Gladiator (2000)': 6.7430014096773663,
 "527: Schindler's List (1993)": 7.1367201337271196,
 '541: Blade Runner (1982)': 7.0010547324841763,
 '593: Silence of the Lambs,

## Normalized

In [11]:
nndfcos = ndfcos.where(ndfcos > 0, 0) # remove similaridades negativas

    What is the ID of the third highest recommended movie for user 5277?

    1259: Stand by Me (1986)  

In [12]:
numerator = ndf.loc[5277].dot(nndfcos)
denominator = df.loc[5277].notnull().dot(nndfcos)

(numerator / denominator).nlargest(5)

260: Star Wars: Episode IV - A New Hope (1977)    1.796591
527: Schindler's List (1993)                      1.793252
1259: Stand by Me (1986)                          0.832911
2396: Shakespeare in Love (1998)                  0.528708
593: Silence of the Lambs, The (1991)             0.451659
Name: 5277, dtype: float64

In [17]:
numerator = ndf.loc[3048].dot(nndfcos)
denominator = df.loc[3048].notnull().dot(nndfcos)

(numerator / denominator).nlargest(5)

1210: Star Wars: Episode VI - Return of the Jedi (1983)    1.754406
318: Shawshank Redemption, The (1994)                      1.672978
296: Pulp Fiction (1994)                                   1.580148
34: Babe (1995)                                            1.578320
2916: Total Recall (1990)                                  1.106043
Name: 3048, dtype: float64